### PR TITLE
[X86][LoopVectorize] Add support for arc and hyperbolic trig functions

### LINKED ISSUE #70079: [HLSL] implement `acos` intrinsic
Implement HLSL acos intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-acos

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x] DXIL backend change
2. [x] SPIRV backend change
3. [x] HLSL frontend change
4. Constraint intrinsic change

### LINKED ISSUE #70080: [HLSL] implement `asin` intrinsic
Implement HLSL `asin` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-asin

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x]   DXIL backend change
2. [x]  SPIRV backend change
3. [x]  HLSL frontend change
4. [x] Constraint intrinsic change

### LINKED ISSUE #70081: [HLSL] implement `atan` intrinsic
Implement HLSL `atan` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-atan

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x]  DXIL backend change
2. [x]  SPIRV backend change
3. [x] HLSL frontend change
4. [x] Constraint intrinsic change

### LINKED ISSUE #70083: [HLSL] implement `sinh` intrinsic
Implement HLSL `sinh` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-sinh

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x]  DXIL backend change
2. [x]  SPIRV backend change
3. [x] HLSL frontend change
4. [x] Constraint intrinsic change

### LINKED ISSUE #70084: [HLSL] implement `cosh` intrinsic
Implement HLSL `cosh` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-cosh

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x]  DXIL backend change
2. [x] SPIRV backend change
3. [x] HLSL frontend change
4. [x] Constraint intrinsic change

### LINKED ISSUE #95966: [ HLSL] Implement `tanh` intrinsic
Implement HLSL `tanh` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-tanh

part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Four parts:
1. [x]  DXIL backend change
2. [x]  SPIRV backend change
3. [x] HLSL frontend change
4. Constraint intrinsic change

### PR DESCRIPTION
**Pull request by farzonl**
This change is part 2 x86 Loop Vectorization of : https://github.com/llvm/llvm-project/pull/96222

It also has veclib call loop vectorization hence the test cases in `llvm/test/Transforms/LoopVectorize/X86/veclib-calls.ll`

finally the last pr missed tests for `llvm/test/CodeGen/X86/fp-strict-libcalls-msvc32.ll` and  `llvm/test/CodeGen/X86/vec-libcalls.ll` so added those aswell.

No evidence was found for arc and  hyperbolic trig  glibc vector  math functions
https://github.com/lattera/glibc/blob/master/sysdeps/x86/fpu/bits/math-vector.h
so no  new `_ZGVbN2v_*` and  `_ZGVdN4v_*` .
So no new tests in `llvm/test/Transforms/LoopVectorize/X86/libm-vector-calls-VF2-VF8.ll`

Also no new svml  and no new tests to: `llvm/test/Transforms/LoopVectorize/X86/svml-calls.ll`
There was  not enough evidence that there were svml arc and hyperbolic trig vector implementations, Documentation was scarces so looked at test cases in [numpy](https://github.com/numpy/SVML/blob/32bf2a98420762a63ab418aaa0a7d6e17eb9627a/linux/avx512/svml_z0_acos_d_la.s#L8).  Someone with more experience with svml should investigate.

## Note 
amd libm doesn't have a vector hyperbolic sine api hence why youi might notice there are no tests for `sinh`.

## History
This change is  part of https://github.com/llvm/llvm-project/issues/87367's investigation on supporting IEEE math operations as intrinsics.
Which was discussed in this RFC:
https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

This change adds  loop vectorization for `acos`, `asin`, `atan`, `cosh`, `sinh`, and `tanh`.
resolves #70079
resolves #70080
resolves #70081
resolves #70083
resolves #70084
resolves #95966

### CODE REVIEW

File: llvm/include/llvm/Analysis/VecFuncs.def [hunk 1]
```diff
@@ -1302,11 +1320,28 @@ TLI_DEFINE_VECFUNC("atanf", "amd_vrs4_atanf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
...
 TLI_DEFINE_VECFUNC("atanf", "amd_vrs16_atanf", FIXED(16), NOMASK, "_ZGV_LLVM_N16v")
 
+TLI_DEFINE_VECFUNC("llvm.atan.f64", "amd_vrd2_atan", FIXED(2), NOMASK, "_ZGV_LLVM_N2v")
+TLI_DEFINE_VECFUNC("llvm.atan.f64", "amd_vrd4_atan", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
+TLI_DEFINE_VECFUNC("llvm.atan.f64", "amd_vrd8_atan", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
+TLI_DEFINE_VECFUNC("llvm.atan.f32", "amd_vrs4_atanf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
+TLI_DEFINE_VECFUNC("llvm.atan.f32", "amd_vrs8_atanf", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
+TLI_DEFINE_VECFUNC("llvm.atan.f32", "amd_vrs16_atanf", FIXED(16), NOMASK, "_ZGV_LLVM_N16v")
+
+TLI_DEFINE_VECFUNC("cosh",  "amd_vrd2_cosh" , FIXED(2), NOMASK, "_ZGV_LLVM_N2v")
```
Comment (farzonl) [inline llvm/include/llvm/Analysis/VecFuncs.def:1330]: @rohitaggarwal007 git blame shows you added the amd vector lib support: https://github.com/llvm/llvm-project/pull/78560.

I wanted to make sure the exclusion of the three libcalls I added was not intentional.

**Comment by farzonl** [comment]
@RKSimon and @topperc If either of you have time could you review this PR? Thanks!

File: llvm/include/llvm/Analysis/VecFuncs.def [hunk 2]
```diff
@@ -1292,8 +1298,20 @@ TLI_DEFINE_VECFUNC("asinf", "amd_vrs4_asinf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
...
 
+TLI_DEFINE_VECFUNC("llvm.asin.f64", "amd_vrd8_asin", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
+TLI_DEFINE_VECFUNC("llvm.asin.f32", "amd_vrs4_asinf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
+TLI_DEFINE_VECFUNC("llvm.asin.f32", "amd_vrs8_asinf", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
+TLI_DEFINE_VECFUNC("llvm.asin.f32", "amd_vrs16_asinf", FIXED(16), NOMASK, "_ZGV_LLVM_N16v")
+
+
 TLI_DEFINE_VECFUNC("acosf", "amd_vrs4_acosf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
 TLI_DEFINE_VECFUNC("acosf", "amd_vrs8_acosf", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
+TLI_DEFINE_VECFUNC("acosf", "amd_vrs16_acosf", FIXED(16), NOMASK, "_ZGV_LLVM_N16v")
```
Comment (topperc) [inline llvm/include/llvm/Analysis/VecFuncs.def:1309]: This line seems not directly related to this patch. Can we do it in a separate patch? Same with the other 2 functions that are not intrinsic related. Hopefully that leads to a more straight forward review from the right owner.

File: llvm/include/llvm/Analysis/VecFuncs.def (hunk 1 above)
Comment (farzonl) [reply llvm/include/llvm/Analysis/VecFuncs.def]: @rohitaggarwal007  Nevermind I'm going to put these in a seperate pr.

File: llvm/include/llvm/Analysis/VecFuncs.def (hunk 2 above)
Comment (farzonl) [reply llvm/include/llvm/Analysis/VecFuncs.def]: Removed, will put up a follow on pr after this one merges.

File: llvm/test/Transforms/LoopVectorize/X86/veclib-calls.ll [hunk 3]
```diff
@@ -431,6 +431,7 @@ for.end:                                          ; preds = %for.body, %entry
   ret void
 }
 
+
```
Comment (topperc) [inline llvm/test/Transforms/LoopVectorize/X86/veclib-calls.ll:434]: Stray blank line change

**Review by topperc — APPROVED** [review]
LGTM
